Add UI config flow to Transport NSW integration

craibo opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add the ability to configure the Transport NSW integration via the UI config flow

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

craibo committed:
Test commit without GPG signing

craibo committed:
Test GPG signed commit

craibo committed:
Revert devcontainer changes

craibo committed:
Enhance Transport NSW config flow with comprehensive options

  - Add route and destination filters to initial config flow
  - Enable name configuration in options flow
  - Fix deprecated config_entry assignment in options flow
  - Add API key setup instructions with direct link to Transport NSW Open Data portal
  - Allow all optional fields (name, route, destination) in both config and options flows
  - Ensure name changes in options properly update config entry data

craibo committed:
Fix logging bug

craibo committed:
Fix logger initialization and remove quality scale from manifest

craibo committed:
Update translations to match config flow changes and add API key instructions

craibo committed:
Add legacy quality scale to manifest

craibo committed:
Revert "Add legacy quality scale to manifest"

This reverts commit 21e2c12cbfc9763a7094f55d8eb3d8359e4eb745.

craibo committed:
Apply code formatting and fix pylint issues for transport_nsw

- Run ruff and prettier formatting on all files
- Remove unused config_validation import
- Add pylint disable for acceptable broad exception catching
- Implement missing is_matching method in config flow
- Remove brand name title to fix validation error
- Update translations to match strings.json structure

craibo committed:
Fix ruff TRY301 issues by abstracting raise statements

craibo committed:
Fix mypy and pylint issues to achieve perfect scores

NoRi2909 reviewed (commented):
A few comments from me as a translator. Although this will probably not be used in any other language we get all strings presented in Lokalise anyway.

craibo committed:
Update homeassistant/components/transport_nsw/strings.json

Co-authored-by: Norbert Rittel <norbert@rittel.de>

craibo committed:
Update homeassistant/components/transport_nsw/strings.json

Co-authored-by: Norbert Rittel <norbert@rittel.de>

craibo committed:
Update homeassistant/components/transport_nsw/strings.json

Co-authored-by: Norbert Rittel <norbert@rittel.de>

craibo committed:
Update homeassistant/components/transport_nsw/strings.json

Co-authored-by: Norbert Rittel <norbert@rittel.de>

craibo committed:
Update homeassistant/components/transport_nsw/strings.json

Co-authored-by: Norbert Rittel <norbert@rittel.de>

craibo committed:
Update homeassistant/components/transport_nsw/strings.json

Co-authored-by: Norbert Rittel <norbert@rittel.de>

craibo committed:
Fix pylint issues: update callback type and move coordinator

craibo committed:
Merge branch 'transport-nsw-add-ui-flow' of https://github.com/craibo/home-assistant-core into transport-nsw-add-ui-flow

craibo committed:
Fix missing Any import in sensor.py

craibo committed:
Apply ruff formatting fixes

craibo committed:
Update generated files for transport_nsw integration

- Add transport_nsw to config_flows.py
- Update transport_nsw entry in integrations.json with config_flow: true
- Change integration_type from hub to service

craibo committed:
Update transport_nsw integration_type to hub

Change integration_type from 'service' to 'hub' in manifest.json to better reflect that this integration manages multiple transport entities from NSW transport services.

craibo committed:
PR review feedback updates and fix test cases

craibo committed:
Add subentry support for Transport NSW integration

- Introduced subentry flow for adding individual transport stops.
- Updated configuration schemas to accommodate subentry details.
- Enhanced validation functions to support subentry input.
- Modified coordinator and sensor classes to handle.

craibo committed:
Merge branch 'dev' into transport-nsw-add-ui-flow

craibo committed:
Remove unused parametrize decorators from config flow tests

craibo committed:
Add comprehensive tests for Transport NSW integration

NoRi2909 reviewed (commented):
A few more points to optimize the strings.

craibo committed:
Update flows and strings

craibo committed:
Update entity flow options

craibo committed:
Merge branch 'dev' into transport-nsw-add-ui-flow

craibo committed:
Add auto reload when a subentry is added or reconfigured

craibo committed:
Fix strings reference

joostlek commented:
So last friday i worked on another PR, if you check the open Nederlandse Spoorwegen PR you can see what kind of is expected for this config flow with subentries PR. But i haven't checked this PR yet, so I'm not sure how much it aligns, but maybe you see a few things that you're missing.

Of course, the Nederlandse Spoorwegen integration configures a bit different, but the goal/destination is the same :)

joostlek committed:
Merge branch 'dev' into transport-nsw-add-ui-flow

craibo commented:
> So looking at this PR, I would say that we shouldn't add the coordinator and change the entity class itself. In theory that code itself can be reused once we migrate to the coordinator, so store it somewhere or make sure you don't squash and push so you don't lose the commits :)

Hi @joostlek 
Apologies for the delayed response. I'm unclear on the what changes you are asking for here?

I've created https://github.com/craibo/ha-transport-nsw so I could start using these updates while waiting on this PR to be merged and they are working as expected.

craibo committed:
Merge branch 'dev' into transport-nsw-add-ui-flow

craibo committed:
Merge branch 'dev' into transport-nsw-add-ui-flow

craibo committed:
Merge branch 'dev' into transport-nsw-add-ui-flow

craibo committed:
Merge branch 'dev' into transport-nsw-add-ui-flow

joostlek reviewed (changes requested):
Okay so the code currently expects either a subentry or a legacy entry, but I am not sure what a legacy entry is as there were no entries before this one. So I would generally recommend now to take a look at https://github.com/home-assistant/core/pull/151567/files. The only thing we do in this PR is leave the sensor platform alone, apart from that the source of the configuration is different.

I would love to see this PR merged, so feel free to message me if you need help, happy to help where needed :)
